Minor OperatorName cleanup

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35058 C++ smoketests for the API.
* #34668 Python-driven testing for the new operator registration API
* **#34620 Minor OperatorName cleanup**
* #34618 Improved error checking
* #34601 Implement schema matching inside Dispatcher
* #34456 Finish implementing schema-less impl, switch all internal registrations to it
* #34455 Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)
* #34221 Relax internal invariants of Dispatcher in preparation for schema-less registration

- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20419614](https://our.internmc.facebook.com/intern/diff/D20419614)

ezyang committed:
Minor OperatorName cleanup

- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Minor OperatorName cleanup"


- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Minor OperatorName cleanup"


- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Minor OperatorName cleanup"


- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20419614](https://our.internmc.facebook.com/intern/diff/D20419614)

[ghstack-poisoned]

ezyang committed:
Update on "Minor OperatorName cleanup"


- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20419614](https://our.internmc.facebook.com/intern/diff/D20419614)

[ghstack-poisoned]

ezyang committed:
Update on "Minor OperatorName cleanup"


- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20419614](https://our.internmc.facebook.com/intern/diff/D20419614)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -196,7 +195,7 @@ class DispatchTable final {
   }
 
-  const std::string& operatorName() const {
+  OperatorName operatorName() const {
```
`const OperatorName&` to avoid the copy?

smessmer commented on aten/src/ATen/core/operator_name.cpp:
```diff
@@ -0,0 +1,17 @@
+}
+
+std::ostream& operator<<(std::ostream& os, const OperatorName& opName) {
+  return os << toString(opName);
```
I'd rather implement it the other way round - have `operator<<` do the actual work and `toString` call `operator<<` with a `std::ostringstream`. In the current implementation, `operator<<` needs to materialize the temporary string. If you do it the other way round, `operator<<` does not need to materialize it anymore an `toString` isn't slower either.

ezyang committed:
Update on "Minor OperatorName cleanup"


- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20419614](https://our.internmc.facebook.com/intern/diff/D20419614)

[ghstack-poisoned]

ezyang committed:
Update on "Minor OperatorName cleanup"


- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20419614](https://our.internmc.facebook.com/intern/diff/D20419614)

[ghstack-poisoned]

ezyang commented:
All CR comments resolved in my local PR

ezyang committed:
Update on "Minor OperatorName cleanup"


- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20419614](https://our.internmc.facebook.com/intern/diff/D20419614)

[ghstack-poisoned]

ezyang committed:
Update on "Minor OperatorName cleanup"


- Store OperatorName instead of string in DispatchTable
- Define operator<< on OperatorName, remove toString calls
- Move toString and operator<< implementations to cpp file

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20419614](https://our.internmc.facebook.com/intern/diff/D20419614)

[ghstack-poisoned]
